xxhash: support SVE by intrinsic code

hzhuang1 opened the pull request:
xxhash: add ARM SVE intrinsic implementation
    
    Implement intrinsic code for ARM SVE. In this patch, it seems that there
    is no improvement. Further optimization will be contained in the later
    patch set.
    
    SCALAR implementation (default)
     ===  benchmarking 4 hash functions  ===
    benchmarking large inputs : from 512 bytes (log9) to 128 MB (log27)
    xxh3   ,  2718,  3093,  3151,  3187,  3217,  3232,  3238,  3223,  3190,  3240,  3241,  3222,  3238,  3049,  2253,  2113,  2102,  2117,  2132
    XXH32  ,  1499,  1520,  1538,  1543,  1543,  1545,  1547,  1535,  1486,  1487,  1484,  1468,  1484,  1435,  1231,  1183,  1184,  1185,  1187
    XXH64  ,  2632,  2874,  3017,  3092,  3132,  3143,  3155,  3122,  2990,  2994,  2998,  3000,  3006,  2842,  2121,  1838,  1833,  1811,  1839
    XXH128 ,  2435,  2872,  3024,  3121,  3187,  3219,  3234,  3222,  3221,  3225,  3225,  3227,  3227,  3034,  1962,  1789,  1866,  1871,  1874
    
    Intrinsic SVE512 implementation
     ===  benchmarking 4 hash functions  ===
    benchmarking large inputs : from 512 bytes (log9) to 128 MB (log27)
    xxh3   ,  1903,  2313,  2469,  2568,  2631,  2663,  2683,  2662,  2673,  2678,  2679,  2678,  2676,  2584,  2236,  2173,  2173,  2163,  2183
    XXH32  ,  1326,  1436,  1495,  1523,  1535,  1543,  1547,  1536,  1505,  1506,  1504,  1507,  1508,  1446,  1246,  1193,  1194,  1194,  1167
    XXH64  ,  2510,  2802,  2977,  3071,  3121,  3136,  3156,  3126,  3041,  3047,  3050,  3050,  3048,  2890,  2043,  1944,  1944,  1945,  1955
    XXH128 ,  1867,  2295,  2462,  2542,  2627,  2663,  2679,  2656,  2661,  2675,  2673,  2647,  2671,  2574,  2218,  2144,  2169,  2175,  2181

Signed-off-by: Haojian Zhuang <haojian.zhuang@linaro.org>
Signed-off-by: Devin Hussey <easyaspi314@users.noreply.github.com>

hzhuang1 committed:
xxhash: add ARM SVE intrinsic implementation

Implement intrinsic code for ARM SVE. In this patch, it seems that there
is no improvement. Further optimization will be contained in the later
patch set.

SCALAR implementation (default)
 ===  benchmarking 4 hash functions  ===
benchmarking large inputs : from 512 bytes (log9) to 128 MB (log27)
xxh3   ,  2718,  3093,  3151,  3187,  3217,  3232,  3238,  3223,  3190,  3240,  3241,  3222,  3238,  3049,  2253,  2113,  2102,  2117,  2132
XXH32  ,  1499,  1520,  1538,  1543,  1543,  1545,  1547,  1535,  1486,  1487,  1484,  1468,  1484,  1435,  1231,  1183,  1184,  1185,  1187
XXH64  ,  2632,  2874,  3017,  3092,  3132,  3143,  3155,  3122,  2990,  2994,  2998,  3000,  3006,  2842,  2121,  1838,  1833,  1811,  1839
XXH128 ,  2435,  2872,  3024,  3121,  3187,  3219,  3234,  3222,  3221,  3225,  3225,  3227,  3227,  3034,  1962,  1789,  1866,  1871,  1874

Intrinsic SVE512 implementation
 ===  benchmarking 4 hash functions  ===
benchmarking large inputs : from 512 bytes (log9) to 128 MB (log27)
xxh3   ,  1903,  2313,  2469,  2568,  2631,  2663,  2683,  2662,  2673,  2678,  2679,  2678,  2676,  2584,  2236,  2173,  2173,  2163,  2183
XXH32  ,  1326,  1436,  1495,  1523,  1535,  1543,  1547,  1536,  1505,  1506,  1504,  1507,  1508,  1446,  1246,  1193,  1194,  1194,  1167
XXH64  ,  2510,  2802,  2977,  3071,  3121,  3136,  3156,  3126,  3041,  3047,  3050,  3050,  3048,  2890,  2043,  1944,  1944,  1945,  1955
XXH128 ,  1867,  2295,  2462,  2542,  2627,  2663,  2679,  2656,  2661,  2675,  2673,  2647,  2671,  2574,  2218,  2144,  2169,  2175,  2181

Signed-off-by: Haojian Zhuang <haojian.zhuang@linaro.org>
Signed-off-by: Devin Hussey <easyaspi314@users.noreply.github.com>

hzhuang1 commented:
@Cyan4973 Could you help to share any comments on this pull request? There's only SVE intrinsic code in this pull request.

Cyan4973 commented:
Hi @hzhuang1 ,
the SVE implementation looks good to me.

I'm essentially concerned about how it could be continuously tested.
Any chance this could run on github action using `qemu` ?

t-mat commented:
> Any chance this could run on github action using qemu ?

It seems `qemu` has some options for SVE.  (Since I don't have environment for now, I just take a quick look.)

- https://github.com/qemu/qemu/blob/master/docs/system/arm/cpu-features.rst#sve-cpu-properties
- https://github.com/qemu/qemu/blob/master/docs/system/arm/cpu-features.rst#sve-cpu-property-examples

Perhaps `-cpu max,sve=on,sve512=on` may work.

Cyan4973 commented on xxhash.h:
```diff
@@ -3197,12 +3209,16 @@ enum XXH_VECTOR_TYPE /* fake enum */ {
 #  elif XXH_VECTOR == XXH_AVX512  /* avx512 */
 #     define XXH_ACC_ALIGN 64
+#  elif XXH_VECTOR == XXH_SVE   /* sve */
+#     define XXH_ACC_ALIGN 64
```
Quick question : 

This variant seems to directly target 512-bit width.

My understanding of `SVE` is that it's more flexible with regards to vector width, compared to Intel's `AVX`.
That being said, the `qemu` documentation implies that there are multiple hardware implementations of `SVE` to expect, from 128 to 512 bits.

Will this vector extension work with all of them, or only specifically on the 512-bit one ?

hzhuang1 replied:
Yes, you're right. SVE ranges from 128-bit to 2048-bit. Should I create a macro to calculate it for SVE?

hzhuang1 replied:
I've configured CI to run code from SVE128 to SVE2048. All of them could work with this XXH_ACC_ALIGN.

Cyan4973 commented on xxhash.h:
```diff
@@ -3178,6 +3184,12 @@ enum XXH_VECTOR_TYPE /* fake enum */ {
 #endif
 
+/* __ARM_FEATURE_SVE is only supported by GCC & Clang. */
+#if (XXH_VECTOR == XXH_SVE) && !defined(__ARM_FEATURE_SVE)
```
Note : maybe minor, but surely worth mentioning.

This transparent fallback method
will make compilation succeed when `SVE` is selected, but actually not present on compilation target.

But it also means that a developer may think that the library is compiled with `SVE` enabled, because it's instructed to, but actually it's not and uses `SCALAR` instead, and the developer would be none the wiser.

I'm wondering if it would be preferable to let compilation fail instead.

Intermediate position, maybe a warning (though unfortunately, I believe `#warning` is not part of `C90` syntax, and `#pragma warning` depends on compiler support. But then, so does `__ARM_FEATURE_SVE`).

hzhuang1 replied:
OK. I'll update it.

hzhuang1 commented:
> Hi @hzhuang1 , the SVE implementation looks good to me.
> 
> I'm essentially concerned about how it could be continuously tested. Any chance this could run on github action using `qemu` ?

SVE could be tested in QEMU. I'm using QEMU to test code. And I only test performance on real platform.

But I don't know how to configure QEMU for SVE in your CI script.

hzhuang1 commented:
> > Any chance this could run on github action using qemu ?
> 
> It seems `qemu` has some options for SVE. (Since I don't have environment for now, I just take a quick look.)
> 
> * https://github.com/qemu/qemu/blob/master/docs/system/arm/cpu-features.rst#sve-cpu-properties
> * https://github.com/qemu/qemu/blob/master/docs/system/arm/cpu-features.rst#sve-cpu-property-examples
> 
> Perhaps `-cpu max,sve=on,sve512=on` may work.

Yes, you're right.

I'm using the config `-cpu max,sve128=on,sve256=on,sve512=on,sve1024=off,sve2048=off` for my test.

t-mat commented:
> I'm using the config -cpu max,sve128=on,sve256=on,sve512=on,sve1024=off,sve2048=off for my test.

Thanks for the info!

> > I'm essentially concerned about how it could be continuously tested. Any chance this could run on github action using qemu ?
>
> But I don't know how to configure QEMU for SVE in your CI script.

You can add your aarch64 SVE test to the following lines

https://github.com/Cyan4973/xxHash/blob/267ba9485bdf3ed200f5cccdfea7d9aa6935eaed/.github/workflows/ci.yml#L381-L385

For example:

```
-   - name: ARM64 (XXH_VECTOR=[ scalar, NEON ])
+   - name: ARM64 (XXH_VECTOR=[ scalar, NEON, SVE ])
      if: ${{ matrix.name == 'ARM64' }}
      run: |
        CPPFLAGS="-DXXH_VECTOR=XXH_SCALAR" LDFLAGS="-static" CC=$XCC RUN_ENV=$XEMU make clean check
        CPPFLAGS="-DXXH_VECTOR=XXH_NEON" LDFLAGS="-static" CC=aarch64-linux-gnu-gcc RUN_ENV=qemu-aarch64-static make clean check
+       CPPFLAGS="-DXXH_VECTOR=XXH_SVE" LDFLAGS="-static" CC=aarch64-linux-gnu-gcc RUN_ENV="qemu-aarch64-static -cpu max,sve128=on,sve256=on,sve512=on,sve1024=off,sve2048=off" make clean check
```

Note that `RUN_ENV` is passed as a "prefix" of the `make check` test commands.

https://github.com/Cyan4973/xxHash/blob/267ba9485bdf3ed200f5cccdfea7d9aa6935eaed/Makefile#L189-L213

hzhuang1 committed:
CI: support ARM64 SVE

Since the vector size of ARM64 SVE ranges from 128-bit to 2048-bit,
configure CI to run with each vector size.

Signed-off-by: Haojian Zhuang <haojian.zhuang@linaro.org>

Cyan4973 reviewed (approved):
Thanks @hzhuang1 , this looks good to me

Cyan4973 commented:
Maybe I was too fast here : I don't see Github Action test results in the test summary.
Seems only Appveyor (Windows) are reported.
